File: gomoku_ai_rust/src/game_types.rs
// ============================================
// Game types and constants — Exact match of game_types.hpp/cpp
// ============================================

pub const BOARD_SIZE: usize = 19;
pub const BOARD_CENTER: i32 = 9;
pub const PLAYER1: i32 = 1; // Human
pub const PLAYER2: i32 = 2; // AI
pub const EMPTY: i32 = 0;
pub const WIN_CAPTURES_NORMAL: i32 = 10;

/// 4 main directions (horizontal, vertical, 2 diagonals)
pub const MAIN_DIRECTIONS: [(i32, i32); 4] = [
    (0, 1),   // Horizontal →
    (1, 0),   // Vertical ↓
    (1, 1),   // Diagonal ↘
    (1, -1),  // Diagonal ↗
];

/// All 8 directions
pub const ALL_DIRECTIONS: [(i32, i32); 8] = [
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),           (0, 1),
    (1, -1),  (1, 0),  (1, 1),
];

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct Move {
    pub x: i32,
    pub y: i32,
}

impl Move {
    pub fn new(x: i32, y: i32) -> Self {
        Move { x, y }
    }

    pub fn invalid() -> Self {
        Move { x: -1, y: -1 }
    }

    pub fn is_valid(&self) -> bool {
        self.x >= 0 && self.y >= 0
            && self.x < BOARD_SIZE as i32
            && self.y < BOARD_SIZE as i32
    }
}

#[derive(Clone)]
pub struct GameState {
    pub board: [[i32; BOARD_SIZE]; BOARD_SIZE],
    pub current_player: i32,
    pub turn_count: i32,
    pub captures: [i32; 2], // [PLAYER1_captures, PLAYER2_captures]
    pub last_human_move: Move,
    pub zobrist_hash: u64,
}

impl GameState {
    pub fn new() -> Self {
        GameState {
            board: [[EMPTY; BOARD_SIZE]; BOARD_SIZE],
            current_player: PLAYER1,
            turn_count: 0,
            captures: [0, 0],
            last_human_move: Move::invalid(),
            zobrist_hash: 0,
        }
    }

    #[inline]
    pub fn is_valid(&self, x: i32, y: i32) -> bool {
        x >= 0 && y >= 0 && x < BOARD_SIZE as i32 && y < BOARD_SIZE as i32
    }

    #[inline]
    pub fn is_empty(&self, x: i32, y: i32) -> bool {
        self.is_valid(x, y) && self.board[x as usize][y as usize] == EMPTY
    }

    #[inline]
    pub fn get_piece(&self, x: i32, y: i32) -> i32 {
        if !self.is_valid(x, y) {
            -1
        } else {
            self.board[x as usize][y as usize]
        }
    }

    #[inline]
    pub fn get_opponent(&self, player: i32) -> i32 {
        if player == PLAYER1 { PLAYER2 } else { PLAYER1 }
    }
}

File: src/core/game_types.cpp
#include "../../include/core/game_types.hpp"
#include "../../include/utils/zobrist_hasher.hpp"
#include <iostream>
#include <cstring>

// Static hasher initialization
const ZobristHasher* GameState::hasher = nullptr;

GameState::GameState() {
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            board[i][j] = EMPTY;
        }
    }
    
    // Initialize last human move as invalid
    lastHumanMove = Move(-1, -1);
    
    // Reset captures and current player
    captures[0] = 0;
    captures[1] = 0;
    currentPlayer = PLAYER1;
    turnCount = 0;
    depth = 0;
    
    // Reset forced capture system
    forcedCaptureMoves.clear();
    forcedCapturePlayer = 0;
    pendingWinPlayer = 0;
    
    // Calculate initial hash if hasher is available
    if (hasher) {
        zobristHash = hasher->computeFullHash(*this);
    }
}

GameState::GameState(const GameState& other) {
    *this = other;
}

GameState& GameState::operator=(const GameState& other) {
    if (this != &other) {
        std::memcpy(board, other.board, sizeof(board));
        captures[0] = other.captures[0];
        captures[1] = other.captures[1];
        currentPlayer = other.currentPlayer;
        turnCount = other.turnCount;
        depth = other.depth;
        zobristHash = other.zobristHash;
        lastHumanMove = other.lastHumanMove;
        
        // Copy forced capture system
        forcedCaptureMoves = other.forcedCaptureMoves;
        forcedCapturePlayer = other.forcedCapturePlayer;
        pendingWinPlayer = other.pendingWinPlayer;
    }
    return *this;
}

bool GameState::isValid(int x, int y) const {
    return x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE;
}

bool GameState::isEmpty(int x, int y) const {
    return isValid(x, y) && board[x][y] == EMPTY;
}

int GameState::getPiece(int x, int y) const {
    return isValid(x, y) ? board[x][y] : -1;
}

int GameState::getOpponent(int player) const {
    return (player == PLAYER1) ? PLAYER2 : PLAYER1;
}

void GameState::initializeHasher() {
    if (!hasher) {
        hasher = new ZobristHasher();
        // Zobrist hasher initialized - logged from main
    }
}

void GameState::cleanupHasher() {
    if (hasher) {
        delete hasher;
        hasher = nullptr;
    }
}

void GameState::updateHashAfterMove(const Move& move, int player, 
                                   const std::vector<Move>& capturedPieces,
                                   int oldCaptures) {
    if (!hasher) {
        std::cerr << "ERROR: Hasher no inicializado. Llama GameState::initializeHasher() primero." << std::endl;
        return;
    }
    
    int playerIndex = player - 1;
    int newCaptures = captures[playerIndex];
    
    zobristHash = hasher->updateHashAfterMove(zobristHash, move, player, 
                                            capturedPieces, oldCaptures, newCaptures);
}

void GameState::recalculateHash() {
    if (!hasher) {
        std::cerr << "ERROR: Hasher no inicializado. Llama GameState::initializeHasher() primero." << std::endl;
        return;
    }
    
    zobristHash = hasher->computeFullHash(*this);
    
    // Hash debug removed to avoid console spam
}
